modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +182,102 @@ private void ensureOpen() {
-                [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
-                retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
-            blobStore.bucket(),
-            blobKey,
-            start + currentOffset,
-            attempt,
-            currentStreamProgress,
-            failuresAfterMeaningfulProgress,
-            maxAttempts
+        logger.log(
+            level,
+            () -> format(
+                """
+                    failed %s [%s/%s] at offset [%s] with purpose [%s]; \
+                    this was attempt [%s] to read this blob which yielded [%s] bytes; in total \
+                    [%s] of the attempts to read this blob have made meaningful progress and do not count towards the maximum number of \
+                    retries; the maximum number of read attempts which do not make meaningful progress is [%s]""",
+                action,
+                blobStore.bucket(),
+                blobKey,
+                start + currentOffset,
+                purpose.getKey(),
+                attempt - 1,
+                currentStreamProgress,
+                failuresAfterMeaningfulProgress,
+                attempt - 1 - failuresAfterMeaningfulProgress
+            ),
+            e
         );
-        if (attempt >= maxAttempts + failuresAfterMeaningfulProgress) {
-            final var finalException = addSuppressedExceptions(e);
-            logger.warn(messageSupplier, finalException);
-            throw finalException;
+    }
+
+    private boolean shouldRetry() {
+        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
+            return false;
         }
-        logger.debug(messageSupplier, e);
-        attempt += 1;
-        if (failures.size() < MAX_SUPPRESSED_EXCEPTIONS) {
-            failures.add(e);
+        if (purpose == OperationPurpose.INDICES) {
+            return true;
+        }
+        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        return attempt < maxAttempts + failuresAfterMeaningfulProgress;
+    }
+
+    private void maybeDelaySafely() {
+        final long delayInMillis = getRetryDelayInMillis();
+        if (delayInMillis > 0) {
+            try {
+                assert purpose == OperationPurpose.INDICES : "only retries for reading indices data should ever reach here";
+                Thread.sleep(delayInMillis);
+            } catch (InterruptedException e) {
+                logger.info("delay interrupted", e);
+            }
+        }
+    }
+
+    // for testing
+    protected long getRetryDelayInMillis() {
+        final long initialDelayInMillis = 10L;
+        final int maxAttempts = blobStore.getMaxRetries() + 1;
+        if (attempt > maxAttempts) {
+            // Cap max delay at 10 * 1024 millis, i.e. it retries every ~10 seconds at a minimum
```

Comment:
nit I keep thinking about the fact that you mentioned we sleep in a blocking mode typically. I wonder if maybe 2 seconds maximum is more suitable than 10 seconds. But I am also not sure why; just a hunch.

Reply:
Changing it to 2 seconds does not really change the blocking nature. The thread just wakes up more often, potentially a bit earlier to pick up if s3/network recovers. As commented [above](https://github.com/elastic/elasticsearch/pull/103300#discussion_r1423368679), I chose 10 seconds after what IndicesService does. If we do get to 10 seconds, the situation is probably _not_ going to resolve very quickly. So waiting a bit longer seems OK.